pkg/progressreader/broadcaster.go

```diff
@@ -0,0 +1,98 @@
+package progressreader
+
+import (
+	"bytes"
+	"io"
+	"sync"
+)
+
+// Broadcaster keeps track of one or more observers watching the progress
+// of an operation. For example, if multiple clients are trying to pull an
+// image, they share a Broadcaster for the download operation.
+type Broadcaster struct {
+	sync.RWMutex
+	// c is a channel that observers block on, waiting for the operation
+	// to finish.
+	c chan struct{}
+	// dataSignals is a list of channels which signal to the observers'
+	// receiveWrites goroutines that new data is available.
+	dataSignals []chan struct{}
+	// history is a buffer of the progress output so far, so a new observer
+	// can catch up.
+	history bytes.Buffer
+}
+
+// NewBroadcaster returns a Broadcaster structure
+func NewBroadcaster() *Broadcaster {
+	return &Broadcaster{
+		c:           make(chan struct{}),
+		dataSignals: []chan struct{}{},
+	}
+}
+
+// receiveWrites runs as a goroutine so that writes don't block the Write
+// function. It writes the new data in broadcaster.history each time there's
+// activity on the dataSignal channel.
+func (broadcaster *Broadcaster) receiveWrites(observer io.Writer, dataSignal <-chan struct{}) {
+	n := 0
+
+	for range dataSignal {
+		broadcaster.RLock()
+		newData := broadcaster.history.Bytes()[n:]
+		// Make a copy of newData so we can release the lock
+		sendData := make([]byte, len(newData), len(newData))
+		copy(sendData, newData)
+		broadcaster.RUnlock()
+
+		written, err := observer.Write(sendData)
+		if err != nil {
+			return
+		}
+		n += written
+	}
+}
+
+// Write adds data to the history buffer, and also writes it to all current
+// observers.
+func (broadcaster *Broadcaster) Write(p []byte) (n int, err error) {
+	broadcaster.Lock()
+	broadcaster.history.Write(p)
+	broadcaster.Unlock()
+
+	// Do a nonblocking write to each dataSignal to cause each writing
+	// goroutine to wake up.
+	for _, dataSignal := range broadcaster.dataSignals {
+		select {
+		case dataSignal <- struct{}{}:
+		}
+	}
+	return len(p), nil
+}
+
+// Add adds an observer to the Broadcaster. The new observer receives the
+// data from the history buffer, and also all subsequent data.
+func (broadcaster *Broadcaster) Add(w io.Writer) {
+	broadcaster.Lock()
+	defer broadcaster.Unlock()
+	dataSignal := make(chan struct{}, 1)
+	broadcaster.dataSignals = append(broadcaster.dataSignals, dataSignal)
+	go broadcaster.receiveWrites(w, dataSignal)
+	select {
+	case dataSignal <- struct{}{}:
+	}
+}
+
+// Close signals to all observers that the operation has finished.
+func (broadcaster *Broadcaster) Close() {
+	broadcaster.Lock()
+	close(broadcaster.c)
```

Comment:
A followup patch that I'm working for the layer pull race on makes it possible to safely call `Close` twice. It shouldn't be an issue for this PR because `Close` is only called once, in `poolRemove`. But I'm happy to include that part of the change in this PR if you'd like. This is the diff:

```go
diff --git a/pkg/progressreader/broadcaster.go b/pkg/progressreader/broadcaster.go
index 088d4a1..4b08ce4 100644
--- a/pkg/progressreader/broadcaster.go
+++ b/pkg/progressreader/broadcaster.go
@@ -23,6 +23,9 @@ type Broadcaster struct {
        history bytes.Buffer
        // wg is a WaitGroup used to wait for all writes to finish on Close
        wg sync.WaitGroup
+       // isClosed is set to true when Close is called to avoid closing c
+       // multiple times.
+       isClosed bool
 }
 
 // NewBroadcaster returns a Broadcaster structure
@@ -124,6 +127,11 @@ func (broadcaster *Broadcaster) Add(w io.Writer) error {
 // Close signals to all observers that the operation has finished.
 func (broadcaster *Broadcaster) Close() {
        broadcaster.Lock()
+       if broadcaster.isClosed {
+               broadcaster.Unlock()
+               return
+       }
+       broadcaster.isClosed = true
        close(broadcaster.c)
        broadcaster.cond.Broadcast()
        broadcaster.Unlock()
```